rest-api/test/unit_test/api_test/questionnaire_response_api_test.py

```diff
@@ -19,20 +20,50 @@ def _questionnaire_response_url(participant_id):
 
 class QuestionnaireResponseApiTest(FlaskTestBase):
 
-  def test_insert(self):
+  def test_insert_raises_400_for_excessively_long_valueString(self):
     participant_id = self.create_participant()
+    questionnaire_id = self.create_questionnaire('questionnaire1.json')
+    url = _questionnaire_response_url(participant_id)
+
+    with open(data_path('questionnaire_response3.json')) as fd:
+      resource = json.load(fd)
+
+    resource['subject']['reference'] = \
+        resource['subject']['reference'].format(participant_id=participant_id)
+
+    resource['questionnaire']['reference'] = \
+        resource['questionnaire']['reference'].format(questionnaire_id=questionnaire_id)
+
+    answer = resource['group']['group'][0]['group'][0]['question'][0]['answer'][0]
+
+    # Check that a string longer than MAX_MYSQL_VARCHAR will not
+    # This one should evaluate to a string that is one char too long; i.e. exactly 64KiB
+    answer['valueString'] = 'a' * (MAX_MYSQL_VARCHAR + 1)
+    self.send_post(url, resource, expected_status=httplib.BAD_REQUEST)
 
+    # Check that a string MAX_MYSQL_VARCHAR will post
+    # This one should be exactly long enough to pass
+    # Remember we need to send the consent first
+    self.send_consent(participant_id)
+    answer['valueString'] = 'a' * MAX_MYSQL_VARCHAR
+    self.send_post(url, resource)
```

Comment:
We should ideally use QuestionnnaireResponseAnswerDao to make sure the row got written out with the full text here.

Reply:
How have you guys been doing that so far?  Writing to the test DB and then reading back to verify, or mocking / monkey-patching the DAO?

Reply:
So most API tests involve writing something with the API and reading it back out with the API.

In this particular case, we're trying to read something -- questionnaire_response_answer row -- that isn't exposed via our API. So we have to use QuestionnaireResponseAnswerDao() here.

Maybe it would be simpler / cleaner to not bother with that here but instead add a test here of the DAO directly:

https://github.com/all-of-us/raw-data-repository/blob/master/rest-api/test/unit_test/dao_test/questionnaire_response_dao_test.py

Reply:
Actually the POST is returning the object as written to the db, I believe, so we can read the response from `send_post` and compare (which I've done).  This call circumvents the DAO's `insert` method so I've thrown a quick check on `from_client_json`, which is the top level DAO method called by the API I think.

Reply:
So it's returning the resource written to the questionnaire_response row, which is good. What it doesn't prove is that we wrote anything out to questionnaire_response_answer. (Though it does prove that we didn't write something out that failed.) So adding something to questionnaire_response_dao_test would be nice.

Reply:
Done.